File: ng2-dart/lib/service/user.service.dart
import 'dart:async';
import 'dart:convert';

import 'package:angular2/core.dart';
import 'package:http/http.dart';

import '../model/user.dart';

@Injectable()
class UserService {

	static final _headers = { 'Content-Type' : 'application/json' };
	static const _restUrl = '/rest/user/';

	final Client _http;

	UserService(this._http);

	Future<List<User>> getUsers() async {
		try {
			final response = await _http.get(_restUrl);
			final users = _extractData(response)
			.map((value) => new User.fromJson(value))
			.toList();

			return users;
		} catch (e) {
			throw _handleError(e);
		}
	}

	Future<User> getUser(int id) async {
		var url = '$_restUrl${id}';
		try {
			final response = await _http.get(url);
			final user = _extractData(response)
			.map((value) => new User.fromJson(value));

			return user;
		} catch (e) {
			throw _handleError(e);
		}
	}

	Future<User> save(User user) async {
		if (user.id != null) {
			return this._put(user);
		} else {
			return this._post(user);
		}
	}	

	Future<User> delete(int id) async {
		var url = '$_restUrl${id}';
		try {
			final response = await _http.delete(url);
			final user = _extractData(response)
			.map((value) => new User.fromJson(value));

			return user;
		} catch (e) {
			throw _handleError(e);
		}
	}

	Future<User> _post(User user) async {
		try {
     		final response = await _http.post(_restUrl,
         	headers: _headers, body: JSON.encode(user));
      
      		return new User.fromJson(_extractData(response));
   		} catch (e) {
      		throw _handleError(e);
    	}
	}

	Future<User> _put(User user) async {
		try {
     		final response = await _http.put(_restUrl,
         	headers: _headers, body: JSON.encode(user));
      
      		return new User.fromJson(_extractData(response));
   		} catch (e) {
      		throw _handleError(e);
    	}
	}

	dynamic _extractData(Response resp) => JSON.decode(resp.body)['data'];

	Exception _handleError(dynamic e) {
    	print(e); // for demo purposes only
    	return new Exception('Server error; cause: $e');
  	}

}

File: ng2-ts/src/app/service/user.service.ts
import { Injectable }    from '@angular/core';
import { Headers, Http } from '@angular/http';

import { Observable } from 'rxjs';

import { User } from '../model/user';

@Injectable()
export class UserService {
    
    private restUrl = '/rest/user/';
    
    constructor(private http: Http) {}
    
    getUsers() {
        return this.http.get(this.restUrl)
            .map( (response:any) => {
                if (response.status == 200) {
                    return response.json() as User[];
                } else {
                    return null;
                }
            });
    }

    getUser(id: number) {
        let url = this.restUrl + id;
        return this.http.get(url)
            .map( (response:any) => {
                if (response.status == 200) {
                    return response.json() as User;
                } else {
                    return null;
                }
            });
    }
    
    save(user: User) {
        if (user.id) {
            return this.put(user);
        }
        return this.post(user);
    }
    
    delete(user: User) {
        let headers = new Headers();
        headers.append('Content-Type', 'application/json');
        let url = this.restUrl + user.id;
        return this.http
            .delete(url, {
                headers: headers
            })
            .map( (response:any) => {
                if (response.status == 200) {
                    return true;
                } else {
                    return false;
                }
            });
    }

    private post(user: User) {
        let headers = new Headers({
            'Content-Type': 'application/json'
        });
        return this.http
            .post(this.restUrl, JSON.stringify(user), {
                headers: headers
            })
            .map( (response:any) => {
                if (response.status == 200) {
                    return response.json() as User;
                } else {
                    return null;
                }
            });
    }

    private put(user: User) {
        let headers = new Headers();
        headers.append('Content-Type', 'application/json');
        let url = this.restUrl;
        return this.http
            .put(url, JSON.stringify(user), {
                headers: headers
            })
            .map( (response:any) => {
                if (response.status == 200) {
                    return response.json() as User;
                } else {
                    return null;
                }
            });
    }
}